pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,231 @@
+				TenantId:        g.config.TenantID,
+				KeyvaultCertUri: certUri,
+				Name:            generateGwListenerCertName(gwObj.Name, listener.Name),
+			}
+			err = buildSPC(spc, spcConf)
+			if err != nil {
+				var userErr userError
+				if errors.As(err, &userErr) {
+					logger.Info("failed to build SecretProviderClass from user error: %s, sending warning event", userErr.userMessage)
+					g.events.Eventf(gwObj, corev1.EventTypeWarning, "InvalidInput", "invalid TLS configuration: %s", userErr.userMessage)
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, fmt.Sprintf("building SPC for listener %s: %s", listener.Name, err.Error()))
+				return ctrl.Result{}, fmt.Errorf("building spc: %w", err)
+			}
+
+			logger.Info(fmt.Sprintf("reconciling SecretProviderClass %s for listener %s", spc.Name, listener.Name))
+			if err = util.Upsert(ctx, g.client, spc); err != nil {
+				fullErr := fmt.Errorf("failed to reconcile SecretProviderClass %s: %w", req.Name, err)
+				logger.Error(err, fullErr.Error())
+				g.events.Event(gwObj, corev1.EventTypeWarning, "FailedUpdateOrCreateSPC", fullErr.Error())
+				return ctrl.Result{}, fullErr
+			}
+
+			logger.Info(fmt.Sprintf("preemptively attaching secret reference for listener %s", listener.Name))
+			newCertRef := gatewayv1.SecretObjectReference{
+				Namespace: to.Ptr(gatewayv1.Namespace(req.Namespace)),
+				Group:     to.Ptr(gatewayv1.Group(corev1.GroupName)),
+				Kind:      to.Ptr(gatewayv1.Kind("Secret")),
+				Name:      gatewayv1.ObjectName(generateGwListenerCertName(gwObj.Name, listener.Name)),
+			}
+			gwObj.Spec.Listeners[index].TLS.CertificateRefs = []gatewayv1.SecretObjectReference{newCertRef}
+			continue
+		}
+		// we should delete the SPC if it exists
+		logger.Info(fmt.Sprintf("attempting to remove unused SPC %s", spc.Name))
+
+		deletionSpc := &secv1.SecretProviderClass{}
+		if err = g.client.Get(ctx, client.ObjectKeyFromObject(spc), deletionSpc); err != nil {
+			if client.IgnoreNotFound(err) != nil {
+				logger.Error(err, fmt.Sprintf("failed to fetch SPC for deletion %s", spc.Name))
+				return ctrl.Result{}, fmt.Errorf("fetching SPC for deletion: %w", err)
+			}
+			continue
+		}
+
+		if manifests.HasTopLevelLabels(deletionSpc.Labels) {
+			// return if we fail to delete, but otherwise, keep going
+			if err = g.client.Delete(ctx, deletionSpc); err != nil {
+				if client.IgnoreNotFound(err) != nil {
+					logger.Error(err, fmt.Sprintf("failed to delete SPC %s", spc.Name))
+					return ctrl.Result{}, fmt.Errorf("deleting SPC: %w", err)
+				}
+				continue
+			}
+		}
+
+	}
+
+	logger.Info("reconciling Gateway resource with new secret refs for each TLS-enabled listener")
+	if err = util.Upsert(ctx, g.client, gwObj); err != nil {
+		fullErr := fmt.Errorf("failed to reconcile Gateway resource %s: %w", req.Name, err)
+		logger.Error(err, fullErr.Error())
+		g.events.Event(gwObj, corev1.EventTypeWarning, "FailedUpdateOrCreateGateway", fullErr.Error())
```

Comment:
This should maybe just be a pure update instead of an upsert.

If a user deletes a gateway we don't want to recreate it

Reply:
yeah good point - didn't realize controller runtime watches for delete events

Reply:
although - if the gateway isn't found we exit early anyway - but I suppose it doesn't hurt in case the delete takes too long to reflect/we somehow end up finding it successfully at the beginning

Reply:
what I'm worried about is the race condition. The gateway is updated then they delete it in the middle but we recreate it here

Reply:
yeah valid - updated to use update